simulation-tests/src/test/java/org/apache/druid/testing/simulate/embedded/EmbeddedDruidServer.java

```diff
@@ -0,0 +1,265 @@
+
+    return dbRule == null
+           ? List.of(referenceHolderModule)
+           : List.of(referenceHolderModule, new TestDerbyModule(dbRule.getConnector()));
+  }
+
+  /**
+   * Creates a JUnit {@link ExternalResource} for this server that can be used
+   * with {@code Rule}, {@code ClassRule} or in a {@code RuleChain}.
+   */
+  ExternalResource junitResource(
+      TemporaryFolder tempDir,
+      EmbeddedZookeeper zk,
+      @Nullable TestDerbyConnector.DerbyConnectorRule dbRule
+  )
+  {
+    return new DruidServerJunitResource(this, tempDir, zk, dbRule);
+  }
+
+  /**
+   * Handler used during initialization of the lifecycle of an embedded server.
+   */
+  interface LifecycleInitHandler
+  {
+    /**
+     * @return Modules that should be used in {@link ServerRunnable#getModules()}.
+     * This list contains modules that cannot be injected into the
+     * {@code StartupInjectorBuilder} as they need dependencies that are only
+     * bound later either in {@link ServerRunnable#getModules()} itself or via
+     * the {@code CoreInjectorBuilder}.
+     */
+    List<? extends Module> getInitModules();
+
+    /**
+     * All implementations of {@link EmbeddedDruidServer} must call this method
+     * from {@link ServerRunnable#initLifecycle(Injector)}.
+     */
+    void onLifecycleInit(Lifecycle lifecycle);
+  }
+
+  /**
+   * Guice module to bind {@link SQLMetadataConnector} to {@link TestDerbyConnector}.
+   * Used in Coordinator and Overlord simulations to connect to an in-memory Derby
+   * database.
+   */
+  private static class TestDerbyModule extends SQLMetadataStorageDruidModule
+  {
+    public static final String TYPE = "derbyInMemory";
+    private final TestDerbyConnector connector;
+
+    public TestDerbyModule(TestDerbyConnector connector)
+    {
+      super(TYPE);
+      this.connector = connector;
+    }
+
+    @Override
+    public void configure(Binder binder)
+    {
+      super.configure(binder);
+
+      binder.bind(MetadataStorage.class).toProvider(NoopMetadataStorageProvider.class);
+
+      PolyBind.optionBinder(binder, Key.get(MetadataStorageProvider.class))
+              .addBinding(TYPE)
+              .to(DerbyMetadataStorageProvider.class)
+              .in(LazySingleton.class);
+
+      PolyBind.optionBinder(binder, Key.get(MetadataStorageConnector.class))
+              .addBinding(TYPE)
+              .toInstance(connector);
+
+      PolyBind.optionBinder(binder, Key.get(SQLMetadataConnector.class))
+              .addBinding(TYPE)
+              .toInstance(connector);
+
+      PolyBind.optionBinder(binder, Key.get(MetadataStorageActionHandlerFactory.class))
+              .addBinding(TYPE)
+              .to(DerbyMetadataStorageActionHandlerFactory.class)
+              .in(LazySingleton.class);
+    }
+  }
+
+  /**
+   * Holder for the service client instances that are being used by this server.
+   */
+  private static class ServiceClientHolder
+  {
+    @Inject
+    CoordinatorClient coordinator;
```

Comment:
I don't really understand why this injection will happen - an instance of it is only present as a private field of this class - could you give a hint? :)

Reply:
Guice takes care of since we have bound it using

`binder.bind(ReferenceHolder.class).toInstance(privateReferenceHolderInstance)`.

Reply:
interesting; I wasn't expecting that during a `toInstance` call that will be interpreted; somehow I was thinking `guice` have to be the one creating the original instance...makes sense ; but I still find it a little bit odd :)
I won't forget this trick :D